doc/designs/quic-design/rx-depacketizer.md

```diff
@@ -0,0 +1,249 @@
+as well.*
+
+It's expected that the QUIC Read Record Layer decrypts the packet and
+parses its packet header, and creates the `OSSL_QUIC_PACKET` structure with
+appropriately populated fields.
+
+The RX depacketizer depends on `packet_data`, `packet_data_length` and
+`packet_payload_offset` to find the packet payload and parse frames out of
+it.  Additionally, it uses `connection`, `packet_number` and `packet_type`.
+
+It's assume that `packet->connection` can be used to find the appropriate
+stream SSL objects attached to the connection.
+
+```c
+struct ossl_quic_packet_st {
+    SSL *connection;
+
+    /*
+     * |packet_data| is the header directly followed by the payload if there
+     * is any.  |packet_header_length| is the length of the packet header,
+     * and |packet_payload_offset| if the offset to the packet payload if
+     * there is any.
+     *
+     * When passing this structure from the TX packetizer to the QUIC Read
+     * Record Layer, |packet_data| is expected to only hold the packet
+     * header, and |frames| holds the rest of the data.  It's up to the QUIC
+     * Read Record Layer to encrypt and send all those chunks of data in a
+     * manner fitting for the network medium with as little copying as
+     * possible.
+     *
+     * When passing this structure to the RX depacketizer from the QUIC Read
+     * Record Layer, |packet_data| is expected to contain the whole packet,
+     * and |frames| to be NULL.
+     */
+    unsigned char *packet_data;
+    size_t packet_data_length;              /* Only used for RX purposes */
+    size_t packet_header_length;
+    size_t packet_payload_offset;           /* Only used for RX purposes */
+
+    STACK_OF(OSSL_QUIC_FRAME) *frames;      /* Only used for TX purposes */
+
+    QUIC_PN packet_number; /* RFC 9000 12.3 */
+
+    /* Packet types */
+    enum {
+        pkt_initial,
+        pkt_handshake,
+        pkt_rtt0,
+        pkt_rtt1,
+        pkt_retry
+    } packet_type;
+
+    /* Spec */
+    unsigned int no_ack : 1;
+    unsigned int no_congestion_control : 1;
+    unsigned int probing : 1;
+    unsigned int flow_controlled : 1;
+};
+
+typedef struct ossl_quic_packet_st OSSL_QUIC_PACKET;
+```
+
+Interactions
+------------
+
+The RX depacketizer receives a packet from an QUIC Read Record Layer, and
+the processes frames in two phases:
+
+1.  [Collect information for the ACK Manager](#collect-information-for-the-ack-manager)
+2.  [Pass frame data](#pass-frame-data)
+
+then loops a second time through the frames to sends the data from each
+frame to diverse other components, depending on the frame type.
+
+### Other components
+
+There are a number of other components that the RX depacketizer wants to
+interact with:
+
+-   [ACK manager]
+-   Handshake manager, which is currently unspecified.  In the [overview],
+    it's called the "TLS Handshake Record Layer", but that may need some
+    wrapping to fit the purpose of a Handshake manager.
```

Comment:
Yes there will need to be something like this. The "TLS Handshake Record layer" is just one piece of the puzzle that will be responsible for taking handshake messages and converting them into frame data. There will need to be something else managing the initiation and lifecycle of the handshake.

Reply:
Thanks for confirming.  I've rephrased a little bit to reflect this more concretely.